### ISSUE TITLE
conftest pull from Gitlab container registry fails

### ISSUE THREAD

**Issue by OmegaVVeapon** [open-policy-agent/conftest#349]
(Opening a ticket on this following some conversation with @garethr on the conftest Slack team)

Conftest push

```
❯ conftest push registry.gitlab.com/omegavveapon/conftest-policies/kubernetes:latest kubernetes
2020/07/22 11:00:45 pushing bundle to: registry.gitlab.com/omegavveapon/conftest-policies/kubernetes:latest
WARN[0000] encountered unknown type application/vnd.cncf.openpolicyagent.policy.layer.v1+rego; children may not be fetched
WARN[0000] reference for unknown type: application/vnd.cncf.openpolicyagent.policy.layer.v1+rego  digest="sha256:af572f7502b324afe155e5c356115cfbcd60da52a2bfa131935974488cdc4641" mediatype=application/vnd.cncf.openpolicyagent.policy.layer.v1+rego size=177
WARN[0000] encountered unknown type application/vnd.cncf.openpolicyagent.policy.layer.v1+rego; children may not be fetched
WARN[0000] reference for unknown type: application/vnd.cncf.openpolicyagent.policy.layer.v1+rego  digest="sha256:e1164ea909407841e65e6686b53a858d6d675e7ab7e365718ed6cb470a4cf646" mediatype=application/vnd.cncf.openpolicyagent.policy.layer.v1+rego size=529
WARN[0000] encountered unknown type application/vnd.cncf.openpolicyagent.policy.layer.v1+rego; children may not be fetched
WARN[0000] encountered unknown type application/vnd.cncf.openpolicyagent.policy.layer.v1+rego; children may not be fetched
WARN[0000] reference for unknown type: application/vnd.cncf.openpolicyagent.policy.layer.v1+rego  digest="sha256:e6ed12612151fc70d544cac1b9b4fb78f740693eb6a9e47ad2fe4483c6927291" mediatype=application/vnd.cncf.openpolicyagent.policy.layer.v1+rego size=667
WARN[0000] encountered unknown type application/vnd.cncf.openpolicyagent.policy.layer.v1+rego; children may not be fetched
WARN[0000] reference for unknown type: application/vnd.cncf.openpolicyagent.policy.layer.v1+rego  digest="sha256:83a28683dff1ab91c46ec9ef52fac1ab2ca224bd086b1e3e03a48ee5cc4ee7d1" mediatype=application/vnd.cncf.openpolicyagent.policy.layer.v1+rego size=107
WARN[0000] encountered unknown type application/vnd.cncf.openpolicyagent.policy.layer.v1+rego; children may not be fetched
WARN[0000] encountered unknown type application/vnd.cncf.openpolicyagent.config.v1+json; children may not be fetched
WARN[0000] reference for unknown type: application/vnd.cncf.openpolicyagent.policy.layer.v1+rego  digest="sha256:fb9c4419361a622430a416c2eb56b52d1dcf6bbad57743d9717354ed1d8a75db" mediatype=application/vnd.cncf.openpolicyagent.policy.layer.v1+rego size=215
WARN[0000] reference for unknown type: application/vnd.cncf.openpolicyagent.policy.layer.v1+rego  digest="sha256:ce44de01af58265129ab84d3ebc6730e89c1f1e440fc57454a29f24f9a0a34a6" mediatype=application/vnd.cncf.openpolicyagent.policy.layer.v1+rego size=1332
WARN[0000] reference for unknown type: application/vnd.cncf.openpolicyagent.config.v1+json  digest="sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a" mediatype=application/vnd.cncf.openpolicyagent.config.v1+json size=2
2020/07/22 11:00:50 pushed bundle with digest: sha256:8570215c558e2e104613ef933c88c58123d05a657703aee5e007843af002075c
```

Conftest pull

```
❯ conftest pull registry.gitlab.com/omegavveapon/conftest-policies/kubernetes:latest
Error: download policies: stat /policy/registry.gitlab.com/omegavveapon/conftest-policies/kubernetes:latest: no such file or directory
```

is it me or the pull thinks that I'm trying to pull from a local file instead of a Docker registry? Is that why it's appending that `/policy` directory to my pull request?

This seems to be confirmed by the fact that when I use the `--policy` flag, it does prepend it to my request.

```
❯ conftest pull --policy foo registry.gitlab.com/omegavveapon/conftest-policies/kubernetes:latest
Error: download policies: stat /foo/registry.gitlab.com/omegavveapon/conftest-policies/kubernetes:latest: no such file or directory
```

I've been able to successfully push AND pull from Azure's Container Registry and a local Docker registry.

**Comment by KeisukeYamashita** [comment]
@OmegaVVeapon Could you try with `oci://registry.gitlab.com/omegavveapon/conftest-policies/kubernetes:latest`, please?

I will raise a PR and try to reproduce soon but for GitLab registries, you need to attach the `oci//` URL scheme to contest to detect that it is trying to pull from OCI registry.

**Comment by OmegaVVeapon** [comment]
Oh wow, thank you! It works!

```
❯ conftest pull oci://registry.gitlab.com/omegavveapon/conftest-policies/kubernetes:latest
WARN[0001] reference for unknown type: application/vnd.cncf.openpolicyagent.policy.layer.v1+rego  digest="sha256:dc373737df73bf3709d925be3266b67dbecdaea3fbe86ce2c5713a214fdf1475" mediatype=application/vnd.cncf.openpolicyagent.policy.layer.v1+rego size=527
WARN[0001] reference for unknown type: application/vnd.cncf.openpolicyagent.policy.layer.v1+rego  digest="sha256:26d1a0234163aeb91e01e98dad711c753b51e1c825ebff5ee5b8d703e4785f54" mediatype=application/vnd.cncf.openpolicyagent.policy.layer.v1+rego size=2408
WARN[0001] blob no name: sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a
WARN[0001] encountered unknown type application/vnd.cncf.openpolicyagent.policy.layer.v1+rego; children may not be fetched
WARN[0001] encountered unknown type application/vnd.cncf.openpolicyagent.policy.layer.v1+rego; children may not be fetched
```

**Comment by KeisukeYamashita** [comment]
This is because the `conftest pull` command only handles  these OCI registries.

* AWS Container Registry
* Google Container Registry
* Azure Container Registry

I've added GitLab Registry support by this PR. After that PR is merged, I will close this issue. Thank you for your confirmation!

**Comment by jpreese** [comment]
Resolved via #351
